### PR TITLE
Remove optional use of statsmodels within EMOS

### PR DESCRIPTION
**Pull request by gavinevans**
Addresses part of #1537 

Description
This PR removes the optional use of statsmodels, meaning that it is now a required module. This aids the further development in #1537 for using a static additional predictor.

Further information in https://github.com/metoppv/mo-blue-team/issues/73#issuecomment-925092996.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

**Comment by tjtg** [comment]
I've created a PR-to-PR which changes this to make statsmodels required only for estimate EMOS functionality. That change avoids a hard dependency on statsmodels for all of IMPROVER but still allows for a lot of the simplification in this PR.

I've had a read through the this PR. All looks OK to me and avoiding two code paths is a nice simplification.

File: improver_tests/calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py [hunk 1]
```diff
@@ -53,32 +53,18 @@
...
-    importlib.import_module("statsmodels")
-except (ModuleNotFoundError, ImportError):
-    STATSMODELS_FOUND = False
-else:
-    STATSMODELS_FOUND = True
-
-
 IGNORED_MESSAGES = [
     "Collapsing a non-contiguous coordinate",  # Originating from Iris
-    "invalid escape sequence",  # Originating from statsmodels
```
Comment (fionaRust) [inline improver_tests/calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py:66]: Why have these been removed?

File: improver_tests/acceptance/SHA256SUMS [hunk 2]
```diff
@@ -100,13 +100,12 @@ c72e48ae88fb3a9c909323376c20d9bcaecec5e1b55a42accdab7201ce8d7cc8  ./estimate-emo
...
+bbb80afc1a2770b3a1a3cb666fe01e7ef7c2e9fc993db0c8688f2563927f8eda  ./estimate-emos-coefficients/normal/realizations/kgo.nc
 42fe8daffb997848086cb08b3d39dd366b561bb1c24fc7ed3e2e54737d772b9a  ./estimate-emos-coefficients/normal/sites/history/20201209T1500Z-PT0012H00M-temperature_at_screen_level.nc
 3cda6a3244db09530585454d323ec13bde6b4cfae24940a8b069469f47aec60c  ./estimate-emos-coefficients/normal/sites/history/20201210T1500Z-PT0012H00M-temperature_at_screen_level.nc
 e5dab98673996a3663093a0353b4efac7910b8176921bd0fbf10aa0f53ae5800  ./estimate-emos-coefficients/normal/sites/history/20201211T1500Z-PT0012H00M-temperature_at_screen_level.nc
 4ff9304b3eca75282ba8aa4a18e0de3404d8510438adccfec7fb20255bfd46c0  ./estimate-emos-coefficients/normal/sites/point_by_point/kgo.nc
-ce9102f00f23256d401eb4cedaa3c70e6a9fcab4dea373ed0f83989215628fa9  ./estimate-emos-coefficients/normal/sites/point_by_point/realizations_kgo.nc
+e9b48f6805af8db049b1b786ee539589ff6186c619a45e9c14cc22a008276f2a  ./estimate-emos-coefficients/normal/sites/point_by_point/realizations_kgo.nc
```
Comment (fionaRust) [inline improver_tests/acceptance/SHA256SUMS:108]: Why has this test changed?

**Review by fionaRust — COMMENTED** [review]
A few questions

File: improver_tests/calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py (hunk 1 above)
Comment (gavinevans) [reply improver_tests/calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py:66]: I've re-added these lines.

File: improver_tests/acceptance/SHA256SUMS (hunk 2 above)
Comment (gavinevans) [reply improver_tests/acceptance/SHA256SUMS:108]: This is because the `test_normal_realizations_point_by_point_sites` acceptance test used to have a `acc.skip_if_statsmodels` decorator, so the output from this test was only generated if statsmodels wasn't present. As statsmodels is always present now, the output from this test has changed due to the use of statsmodels to estimate an initial guess. Previously, as this test was only run without statsmodels, a default initial guess would have been used, which would lead to different values for the coefficients being estimated.

Comment (fionaRust) [reply improver_tests/acceptance/SHA256SUMS:108]: I think this is ok, thanks for investigating

**Review by fionaRust — APPROVED** [review]
Thanks Gavin

File: envs/environment_py37_iris30.yml [hunk 3]
```diff
@@ -21,10 +21,10 @@ dependencies:
   - scipy=1.6
   - sigtools
   - sphinx
-  - statsmodels
   # Optional
   - numba
   - pysteps=1.4.1
+  - statsmodels
```
Comment (MoseleyS) [inline envs/environment_py37_iris30.yml:27]: I think I need this explaining. The statsmodel dependency has moved from "Required" to "Optional", but the PR removes the optional status of this module, which I think makes it "Required".

File: envs/environment_py38_iris30.yml [hunk 4]
```diff
@@ -21,11 +21,7 @@ dependencies:
   - scipy=1.6
   - sigtools
   - sphinx
-  - statsmodels
   # Optional
-  - numba
-  - pysteps=1.4.1
-  - timezonefinder=4.1.0
```
Comment (MoseleyS) [inline envs/environment_py38_iris30.yml:28]: Why are these removed?

File: improver/calibration/ensemble_calibration.py [hunk 5]
```diff
@@ -994,6 +993,8 @@ def compute_initial_guess(
             List of coefficients to be used as initial guess.
             Order of coefficients is [alpha, beta, gamma, delta].
         """
+        import statsmodels.api as sm
```
Comment (MoseleyS) [inline improver/calibration/ensemble_calibration.py:996]: Why is this import statement not at the top of the file with all the other import statements?

**Review by MoseleyS — CHANGES_REQUESTED** [review]
I've got a few questions about this PR, hopefully it is just my lack of understanding.

File: improver_tests/calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py [hunk 6]
```diff
@@ -508,6 +484,7 @@ def setUp(self):
         halo surrounding the original data.
         Set up expected results for different situations.
         """
+        pytest.importorskip("statsmodels")
```
[the remark this answers is not in the document]
Comment (gavinevans) [reply improver_tests/calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py:487]: Thanks @MoseleyS. I think your queries are all related to @tjtg's PR to this PR (https://github.com/gavinevans/improver/pull/13). You're correct that this PR makes statsmodels no longer optional, however, as statsmodels is still only used in one particular place in the codebase, rather than having widespread usage like numpy or Iris, @tjtg has recommended that statsmodels is regarded as an optional module. I think this is a similar set-up to pysteps. Therefore, if someone wants to use components of improver, but isn't interested in EMOS, then they don't need to have statsmodels in their environment.

File: envs/environment_py37_iris30.yml (hunk 3 above)
Comment (gavinevans) [reply envs/environment_py37_iris30.yml:27]: My previous understanding of "Required" and "Optional" was just like you suggest, which is why I made this PR: https://github.com/metoppv/improver/pull/1556. However, the preference, is to treat statsmodels more like pysteps, so even though it is required for EMOS, it is optional for the improver codebase, as someone might want to use improver, but not the EMOS component.

File: envs/environment_py38_iris30.yml (hunk 4 above)
Comment (gavinevans) [reply envs/environment_py38_iris30.yml:28]: There is an interest in representing different things with these different environments with this environment being slightly slimmer. This is the approach chosen by @tjtg in https://github.com/gavinevans/improver/pull/13.

File: improver/calibration/ensemble_calibration.py (hunk 5 above)
Comment (gavinevans) [reply improver/calibration/ensemble_calibration.py:996]: As mentioned above, this is because statsmodels is being treated as optional for the codebase, even though it is required for EMOS. Therefore, if someone isn't using the EMOS component of improver, then they can use a slimmer environment.

File: envs/environment_py38_iris30.yml [hunk 7]
```diff
@@ -21,11 +21,7 @@ dependencies:
   - scipy=1.6
   - sigtools
   - sphinx
-  - statsmodels
```
Comment (MoseleyS) [inline envs/environment_py38_iris30.yml:24]: Hi @tjtg. Can we add a comment to this environment stating what is NOT supported, so that we know what to expect. e.g. 
```
# This environment does not support the following CLIs:
- estimate-emos-coefficients
- generate-timezone-mask-ancillary
- nowcast-accumulate
- nowcast-extrapolate
- nowcast-optical-flow-from-winds
- generate-advection-velocities-from-winds
```

**Review by MoseleyS — APPROVED** [review]
Approved. I will follow up on a comment in the environment as a separate PR when Tom has had a chance to respond to my question.
